extensions/ql-vscode/src/packages/commands/CommandManager.ts

```diff
@@ -1 +1,72 @@
-export class CommandManager {}
+/**
+ * Contains a generic implementation of typed commands.
+ *
+ * This allows different parts of the extension to register commands with a certain type,
+ * and then allow other parts to call those commands in a well-typed manner.
+ */
+
+export interface Disposable {
+  dispose(): void;
+}
+
+/**
+ * A command function is a completely untyped command.
+ */
+export type CommandFunction = (...args: any[]) => Promise<unknown>;
+
+/**
+ * The command manager basically takes a single input, the type
+ * of all the known commands. The second parameter is provided by
+ * default (and should not be needed by the caller) it is a
+ * technicality to allow the type system to look up commands.
+ */
+export class CommandManager<
+  Commands extends Record<string, CommandFunction>,
+  CommandName extends keyof Commands & string = keyof Commands & string,
+> implements Disposable
+{
+  // TODO: should this be a map?
+  // TODO: handle multiple command names
+  private commands: Disposable[] = [];
+
+  constructor(
+    private readonly commandRegister: <T extends CommandName>(
+      commandName: T,
+      definition: Commands[T],
```

Comment:
Is this the "command function". I'm wondering if the terms are the same and whether we could just use a single term.

Reply:
I cannot exactly remember why we needed this, @koesie10 do you have an answer here?

Reply:
Yes, this is the command function. My PoC originally had a way to register commands with progress using an object, so in that case it was really a definition. Now it's just a function, so renaming it to `commandFunction` or even `fn` would make sense to me.

Reply:
Great! One less term 😄